File: src/api/spacer/agentSpaces.ts
/*
 * Copyright (c) 2023-2025 Datalayer, Inc.
 * Distributed under the terms of the Modified BSD License.
 */

/**
 * Agent Spaces API functions for the Datalayer platform.
 *
 * Provides functions for creating, listing, updating, and deleting agent spaces.
 *
 * @module api/spacer/agent-spaces
 */

import { requestDatalayerAPI } from '../DatalayerApi';
import { API_BASE_PATHS, DEFAULT_SERVICE_URLS } from '../constants';

// =============================================================================
// Types
// =============================================================================

/**
 * MCP Server tool configuration
 */
export interface MCPServerTool {
  name: string;
  description: string;
  enabled: boolean;
}

/**
 * MCP Server configuration
 */
export interface MCPServer {
  id: string;
  name: string;
  url?: string;
  enabled: boolean;
  tools: MCPServerTool[];
  command?: string;
  args: string[];
  isAvailable: boolean;
  transport: 'stdio' | 'http';
}

/**
 * Agent skill configuration
 */
export interface AgentSkill {
  id: string;
  name: string;
  description: string;
  version: string;
  tags: string[];
  enabled: boolean;
}

/**
 * Agent specification model
 */
export interface AgentSpec {
  id: string;
  name: string;
  description: string;
  tags: string[];
  enabled: boolean;
  mcpServers: MCPServer[];
  skills: AgentSkill[];
  environmentName: string;
  icon?: string;
  color?: string;
  /** Chat suggestions to show users what this agent can do */
  suggestions?: string[];
}

/**
 * Agent space status
 */
export type AgentSpaceStatus =
  | 'starting'
  | 'running'
  | 'paused'
  | 'terminated'
  | 'archived';

/**
 * Agent space data
 */
export interface AgentSpaceData {
  id: string;
  name: string;
  description: string;
  status: AgentSpaceStatus;
  isPublic: boolean;
  tags: string[];
  messageCount: number;
  createdAt?: string;
  updatedAt?: string;
  thumbnail?: string;
  lastMessage?: string;
  author?: string;
  avatarUrl?: string;
  stars?: number;
  podName?: string;
  runtimeUrl?: string;
  creatorUid?: string;
  creatorHandle?: string;
  agentSpec?: AgentSpec;
}

/**
 * Request to create an agent space
 */
export interface CreateAgentSpaceRequest {
  /** Name of the agent space */
  name: string;
  /** Parent space ID */
  spaceId: string;
  /** Description */
  description?: string;
  /** Tags for categorization */
  tags?: string[];
  /** Initial status */
  status?: AgentSpaceStatus;
  /** Whether publicly visible */
  isPublic?: boolean;
  /** Agent specification */
  agentSpec?: AgentSpec;
  /** Thumbnail URL */
  thumbnail?: string;
}

/**
 * Request to update an agent space
 */
export interface UpdateAgentSpaceRequest {
  name?: string;
  description?: string;
  tags?: string[];
  status?: AgentSpaceStatus;
  isPublic?: boolean;
  agentSpec?: AgentSpec;
  podName?: string;
  runtimeUrl?: string;
  messageCount?: number;
  lastMessage?: string;
  thumbnail?: string;
}

/**
 * Response for single agent space operations
 */
export interface AgentSpaceResponse {
  success: boolean;
  message: string;
  agentSpace?: AgentSpaceData;
}

/**
 * Response for listing agent spaces
 */
export interface AgentSpacesResponse {
  success: boolean;
  message: string;
  agentSpaces?: AgentSpaceData[];
}

// =============================================================================
// API Functions
// =============================================================================

/**
 * Create a new agent space.
 * @param token - Authentication token
 * @param data - Agent space creation configuration
 * @param baseUrl - Base URL for the API (defaults to production)
 * @returns Promise resolving to the created agent space response
 */
export const createAgentSpace = async (
  token: string,
  data: CreateAgentSpaceRequest,
  baseUrl: string = DEFAULT_SERVICE_URLS.SPACER,
): Promise<AgentSpaceResponse> => {
  return requestDatalayerAPI<AgentSpaceResponse>({
    url: `${baseUrl}${API_BASE_PATHS.SPACER}/agent-spaces`,
    method: 'POST',
    token,
    body: data,
  });
};

/**
 * List agent spaces for the current user.
 * @param token - Authentication token
 * @param baseUrl - Base URL for the API (defaults to production)
 * @returns Promise resolving to the list of agent spaces
 */
export const listAgentSpaces = async (
  token: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.SPACER,
): Promise<AgentSpacesResponse> => {
  return requestDatalayerAPI<AgentSpacesResponse>({
    url: `${baseUrl}${API_BASE_PATHS.SPACER}/agent-spaces`,
    method: 'GET',
    token,
  });
};

/**
 * List all public agent spaces (Library).
 * @param baseUrl - Base URL for the API (defaults to production)
 * @returns Promise resolving to the list of public agent spaces
 */
export const listPublicAgentSpaces = async (
  baseUrl: string = DEFAULT_SERVICE_URLS.SPACER,
): Promise<AgentSpacesResponse> => {
  return requestDatalayerAPI<AgentSpacesResponse>({
    url: `${baseUrl}${API_BASE_PATHS.SPACER}/agent-spaces/public`,
    method: 'GET',
  });
};

/**
 * Get an agent space by UID.
 * @param token - Authentication token
 * @param uid - The agent space UID
 * @param baseUrl - Base URL for the API (defaults to production)
 * @returns Promise resolving to the agent space
 */
export const getAgentSpace = async (
  token: string,
  uid: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.SPACER,
): Promise<AgentSpaceResponse> => {
  return requestDatalayerAPI<AgentSpaceResponse>({
    url: `${baseUrl}${API_BASE_PATHS.SPACER}/agent-spaces/${uid}`,
    method: 'GET',
    token,
  });
};

/**
 * Update an agent space.
 * @param token - Authentication token
 * @param uid - The agent space UID
 * @param data - Update data
 * @param baseUrl - Base URL for the API (defaults to production)
 * @returns Promise resolving to the updated agent space
 */
export const updateAgentSpace = async (
  token: string,
  uid: string,
  data: UpdateAgentSpaceRequest,
  baseUrl: string = DEFAULT_SERVICE_URLS.SPACER,
): Promise<AgentSpaceResponse> => {
  return requestDatalayerAPI<AgentSpaceResponse>({
    url: `${baseUrl}${API_BASE_PATHS.SPACER}/agent-spaces/${uid}`,
    method: 'PUT',
    token,
    body: data,
  });
};

/**
 * Delete an agent space.
 * @param token - Authentication token
 * @param uid - The agent space UID
 * @param baseUrl - Base URL for the API (defaults to production)
 * @returns Promise resolving when deleted
 */
export const deleteAgentSpace = async (
  token: string,
  uid: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.SPACER,
): Promise<void> => {
  await requestDatalayerAPI<void>({
    url: `${baseUrl}${API_BASE_PATHS.SPACER}/agent-spaces/${uid}`,
    method: 'DELETE',
    token,
  });
};

/**
 * Make an agent space public (add to Library).
 * @param token - Authentication token
 * @param uid - The agent space UID
 * @param baseUrl - Base URL for the API (defaults to production)
 * @returns Promise resolving to the updated agent space
 */
export const makeAgentSpacePublic = async (
  token: string,
  uid: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.SPACER,
): Promise<AgentSpaceResponse> => {
  return requestDatalayerAPI<AgentSpaceResponse>({
    url: `${baseUrl}${API_BASE_PATHS.SPACER}/agent-spaces/${uid}/public`,
    method: 'POST',
    token,
  });
};

/**
 * Make an agent space private (remove from Library).
 * @param token - Authentication token
 * @param uid - The agent space UID
 * @param baseUrl - Base URL for the API (defaults to production)
 * @returns Promise resolving to the updated agent space
 */
export const makeAgentSpacePrivate = async (
  token: string,
  uid: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.SPACER,
): Promise<AgentSpaceResponse> => {
  return requestDatalayerAPI<AgentSpaceResponse>({
    url: `${baseUrl}${API_BASE_PATHS.SPACER}/agent-spaces/${uid}/private`,
    method: 'POST',
    token,
  });
};

File: datalayer_core/mixins/agent_spaces.py
# Copyright (c) 2023-2025 Datalayer, Inc.
# Distributed under the terms of the Modified BSD License.

"""Agent Spaces mixin for Datalayer Core client."""

import logging
from typing import Any, Optional

logger = logging.getLogger(__name__)


class AgentSpacesMixin:
    """Mixin for agent space operations in Datalayer."""

    def list_agent_spaces(self) -> list[dict[str, Any]]:
        """
        List agent spaces for the current user.

        Returns
        -------
        list[dict[str, Any]]
            List of agent spaces.
        """
        response = self._fetch(  # type: ignore
            "{}/api/spacer/v1/agent-spaces".format(self.urls.spacer_url),  # type: ignore
            method="GET",
        )

        if response.status_code != 200:
            logger.error(f"Failed to list agent spaces: HTTP {response.status_code}")
            return []

        data = response.json()
        if data.get("success"):
            return data.get("agentSpaces", [])
        return []

    def list_public_agent_spaces(self) -> list[dict[str, Any]]:
        """
        List all public agent spaces (Library).

        Returns
        -------
        list[dict[str, Any]]
            List of public agent spaces.
        """
        response = self._fetch(  # type: ignore
            "{}/api/spacer/v1/agent-spaces/public".format(self.urls.spacer_url),  # type: ignore
            method="GET",
        )

        if response.status_code != 200:
            logger.error(
                f"Failed to list public agent spaces: HTTP {response.status_code}"
            )
            return []

        data = response.json()
        if data.get("success"):
            return data.get("agentSpaces", [])
        return []

    def get_agent_space(self, uid: str) -> Optional[dict[str, Any]]:
        """
        Get an agent space by UID.

        Parameters
        ----------
        uid : str
            The agent space UID.

        Returns
        -------
        Optional[dict[str, Any]]
            The agent space data, or None if not found.
        """
        response = self._fetch(  # type: ignore
            "{}/api/spacer/v1/agent-spaces/{}".format(self.urls.spacer_url, uid),  # type: ignore
            method="GET",
        )

        if response.status_code == 404:
            return None

        if response.status_code != 200:
            logger.error(f"Failed to get agent space: HTTP {response.status_code}")
            return None

        data = response.json()
        if data.get("success"):
            return data.get("agentSpace")
        return None

    def create_agent_space(
        self,
        name: str,
        space_id: str,
        description: str = "",
        tags: Optional[list[str]] = None,
        status: str = "paused",
        is_public: bool = False,
        agent_spec: Optional[dict[str, Any]] = None,
        thumbnail: Optional[str] = None,
    ) -> Optional[dict[str, Any]]:
        """
        Create a new agent space.

        Parameters
        ----------
        name : str
            Name of the agent space.
        space_id : str
            Parent space ID.
        description : str
            Description of the agent space.
        tags : Optional[list[str]]
            Tags for categorization.
        status : str
            Initial status (default: 'paused').
        is_public : bool
            Whether publicly visible (default: False).
        agent_spec : Optional[dict[str, Any]]
            Agent specification.
        thumbnail : Optional[str]
            Thumbnail URL.

        Returns
        -------
        Optional[dict[str, Any]]
            The created agent space, or None on failure.
        """
        body = {
            "name": name,
            "spaceId": space_id,
            "description": description,
            "tags": tags or [],
            "status": status,
            "isPublic": is_public,
        }

        if agent_spec:
            body["agentSpec"] = agent_spec
        if thumbnail:
            body["thumbnail"] = thumbnail

        response = self._fetch(  # type: ignore
            "{}/api/spacer/v1/agent-spaces".format(self.urls.spacer_url),  # type: ignore
            method="POST",
            body=body,
        )

        if response.status_code not in (200, 201):
            logger.error(f"Failed to create agent space: HTTP {response.status_code}")
            return None

        data = response.json()
        if data.get("success"):
            return data.get("agentSpace")
        return None

    def update_agent_space(
        self,
        uid: str,
        **kwargs: Any,
    ) -> Optional[dict[str, Any]]:
        """
        Update an agent space.

        Parameters
        ----------
        uid : str
            The agent space UID.
        **kwargs : Any
            Fields to update (name, description, tags, status, isPublic,
            agentSpec, podName, runtimeUrl, messageCount, lastMessage, thumbnail).

        Returns
        -------
        Optional[dict[str, Any]]
            The updated agent space, or None on failure.
        """
        # Build update body from provided kwargs
        body = {}
        valid_fields = {
            "name",
            "description",
            "tags",
            "status",
            "isPublic",
            "agentSpec",
            "podName",
            "runtimeUrl",
            "messageCount",
            "lastMessage",
            "thumbnail",
        }
        for key, value in kwargs.items():
            if key in valid_fields and value is not None:
                body[key] = value

        if not body:
            logger.warning("No valid fields provided for update")
            return self.get_agent_space(uid)

        response = self._fetch(  # type: ignore
            "{}/api/spacer/v1/agent-spaces/{}".format(self.urls.spacer_url, uid),  # type: ignore
            method="PUT",
            body=body,
        )

        if response.status_code == 404:
            return None

        if response.status_code != 200:
            logger.error(f"Failed to update agent space: HTTP {response.status_code}")
            return None

        data = response.json()
        if data.get("success"):
            return data.get("agentSpace")
        return None

    def delete_agent_space(self, uid: str) -> bool:
        """
        Delete an agent space.

        Parameters
        ----------
        uid : str
            The agent space UID.

        Returns
        -------
        bool
            True if deleted, False if not found or on error.
        """
        response = self._fetch(  # type: ignore
            "{}/api/spacer/v1/agent-spaces/{}".format(self.urls.spacer_url, uid),  # type: ignore
            method="DELETE",
        )

        if response.status_code == 204:
            return True

        if response.status_code == 404:
            logger.warning(f"Agent space not found: {uid}")
            return False

        logger.error(f"Failed to delete agent space: HTTP {response.status_code}")
        return False

    def make_agent_space_public(self, uid: str) -> Optional[dict[str, Any]]:
        """
        Make an agent space public (add to Library).

        Parameters
        ----------
        uid : str
            The agent space UID.

        Returns
        -------
        Optional[dict[str, Any]]
            The updated agent space, or None on failure.
        """
        response = self._fetch(  # type: ignore
            "{}/api/spacer/v1/agent-spaces/{}/public".format(self.urls.spacer_url, uid),  # type: ignore
            method="POST",
        )

        if response.status_code == 404:
            return None

        if response.status_code != 200:
            logger.error(
                f"Failed to make agent space public: HTTP {response.status_code}"
            )
            return None

        data = response.json()
        if data.get("success"):
            return data.get("agentSpace")
        return None

    def make_agent_space_private(self, uid: str) -> Optional[dict[str, Any]]:
        """
        Make an agent space private (remove from Library).

        Parameters
        ----------
        uid : str
            The agent space UID.

        Returns
        -------
        Optional[dict[str, Any]]
            The updated agent space, or None on failure.
        """
        response = self._fetch(  # type: ignore
            "{}/api/spacer/v1/agent-spaces/{}/private".format(
                self.urls.spacer_url, uid
            ),  # type: ignore
            method="POST",
        )

        if response.status_code == 404:
            return None

        if response.status_code != 200:
            logger.error(
                f"Failed to make agent space private: HTTP {response.status_code}"
            )
            return None

        data = response.json()
        if data.get("success"):
            return data.get("agentSpace")
        return None
